### PR TITLE
feat: support for uploading and downloading logs/artifact/cache to/from blob storage configured in external cluster 

### PR DESCRIPTION
**Pull request by prakash100198**
Fixes https://github.com/devtron-labs/devtron/issues/4184

This pr aims at introducing support for uploading and downloading logs/artifact/cache to/from blob storage configured in external cluster with some refactoring around blob storage service operations.

### CODE REVIEW

File: helper/CloudUploadHelper.go [hunk 1]
```diff
@@ -56,3 +42,55 @@ func createBlobStorageRequest(cloudProvider blob_storage.BlobStorageType, source
...
 }
+
+func UpdateCloudHelperBaseConfigForExtCluster(cloudHelperBaseConfig *util.CloudHelperBaseConfig) {
+	log.Println(util.DEVTRON, "using external cluster blob")
+	blobStorageConfig, err := GetBlobStorageConfig()
+	if err != nil {
+		log.Println(util.DEVTRON, "error in getting blob storage config, err : ", err)
+	}
+	log.Println(util.DEVTRON, "blob storage config: ", blobStorageConfig)
+	log.Println(util.DEVTRON, "cloudHelperBaseConfig: ", cloudHelperBaseConfig)
```
Comment (kripanshdevtron) [inline helper/CloudUploadHelper.go:53]: Whats the need of logging complete config object containing pass key as well ??

File: helper/CloudUploadHelper.go [hunk 2]
```diff
@@ -56,3 +42,55 @@ func createBlobStorageRequest(cloudProvider blob_storage.BlobStorageType, source
 	}
 	return request
 }
+
+func UpdateCloudHelperBaseConfigForExtCluster(cloudHelperBaseConfig *util.CloudHelperBaseConfig) {
```
Comment (kripanshdevtron) [inline helper/CloudUploadHelper.go:46]: refactoring needed

File: helper/EventHelper.go [hunk 3]
```diff
@@ -47,6 +47,9 @@ const (
 	MANAGED_DOCKERFILE_BUILD_TYPE CiBuildType = "managed-dockerfile-build"
 	BUILD_SKIP_BUILD_TYPE         CiBuildType = "skip-build"
 	BUILDPACK_BUILD_TYPE          CiBuildType = "buildpack-build"
+	BlobStorageObjectTypeCache                = "cache"
```
Comment (kripanshdevtron) [inline helper/EventHelper.go:50]: can refer from `common-lib`

File: helper/EventHelper.go [hunk 4]
```diff
@@ -170,6 +176,58 @@ type CommonWorkflowRequest struct {
...
+		CloudProvider:           c.CloudProvider,
+		UseExternalClusterBlob:  c.UseExternalClusterBlob,
+		BlobStorageS3Config:     c.BlobStorageS3Config,
+		AzureBlobConfig:         c.AzureBlobConfig,
+		GcpBlobConfig:           c.GcpBlobConfig,
+		BlobStorageObjectType:   blobStorageObjectType,
+	}
+}
+
+type BlobStorageConfig struct {
```
Comment (kripanshdevtron) [inline helper/EventHelper.go:192]: Please add some comments explaining why we need it here as well, when it was also coming from orchestrator

File: App.go [hunk 5]
```diff
@@ -142,14 +142,24 @@ func uploadLogs(event helper.CiCdTriggerEvent, exitCode *int) {
...
+	cloudHelperConfig := &util.CloudHelperBaseConfig{
+		StorageModuleConfigured: storageModuleConfigured,
+		BlobStorageLogKey:       blobStorageLogKey,
+		CloudProvider:           cloudProvider,
+		UseExternalClusterBlob:  event.CommonWorkflowRequest.UseExternalClusterBlob,
+		BlobStorageS3Config:     blobStorageS3Config,
+		AzureBlobConfig:         azureBlobConfig,
+		GcpBlobConfig:           gcpBlobConfig,
+		BlobStorageObjectType:   blob_storage.BlobStorageObjectTypeLog,
+	}
```
Comment (subhashish-devtron) [inline App.go:154]: discuss this offline

File: helper/CloudUploadHelper.go (hunk 2 above)
Comment (subhashish-devtron) [reply helper/CloudUploadHelper.go:46]: +1

File: helper/EventHelper.go (hunk 3 above)
Comment (prakash100198) [reply helper/EventHelper.go]: done

File: helper/EventHelper.go (hunk 4 above)
Comment (prakash100198) [reply helper/EventHelper.go:191]: done

**Review by subhashish-devtron — APPROVED** [review]

**Pull request merged**